src/Surfnet/ServiceProviderDashboard/Application/CommandHandler/Service/DeleteServiceCommandHandler.php

```diff
@@ -0,0 +1,118 @@
+ * limitations under the License.
+ */
+
+namespace Surfnet\ServiceProviderDashboard\Application\CommandHandler\Service;
+
+use Exception;
+use League\Tactician\CommandBus;
+use Psr\Log\LoggerInterface;
+use Surfnet\ServiceProviderDashboard\Application\Command\Entity\DeleteCommandFactory;
+use Surfnet\ServiceProviderDashboard\Application\Command\Service\DeleteServiceCommand;
+use Surfnet\ServiceProviderDashboard\Application\CommandHandler\CommandHandler;
+use Surfnet\ServiceProviderDashboard\Application\Service\EntityServiceInterface;
+use Surfnet\ServiceProviderDashboard\Application\ViewObject\EntityList;
+use Surfnet\ServiceProviderDashboard\Domain\Repository\ServiceRepository;
+
+class DeleteServiceCommandHandler implements CommandHandler
+{
+    /**
+     * @var ServiceRepository
+     */
+    private $serviceRepository;
+
+    /**
+     * @var EntityServiceInterface
+     */
+    private $entityService;
+
+    /**
+     * @var DeleteCommandFactory
+     */
+    private $deleteCommandFactory;
+
+    /**
+     * @var CommandBus
+     */
+    private $commandBus;
+
+    /**
+     * @var LoggerInterface
+     */
+    private $logger;
+
+    public function __construct(
+        ServiceRepository $serviceRepository,
+        EntityServiceInterface $entityService,
+        DeleteCommandFactory $deleteCommandFactory,
+        CommandBus $commandBus,
+        LoggerInterface $logger
+    ) {
+        $this->serviceRepository = $serviceRepository;
+        $this->entityService = $entityService;
+        $this->deleteCommandFactory = $deleteCommandFactory;
+        $this->commandBus = $commandBus;
+        $this->logger = $logger;
+    }
+
+    /**
+     * @param DeleteServiceCommand $command
+     */
+    public function handle(DeleteServiceCommand $command)
+    {
+        $serviceId = $command->getId();
+        $service = $this->serviceRepository->findById($serviceId);
+
+        $this->logger->info(sprintf('Removing "%s" and all its entities.', $service->getName()));
+
+        // Remove the entities of the service
+        $entityList = $this->entityService->getEntityListForService($service);
+        $nofEntities = count($entityList->getEntities());
+        if ($nofEntities > 0) {
+            $this->logger->info(sprintf('Removing "%d" entities.', $nofEntities));
+            // Invoke the correct entity delete command on the command bus
+            $this->removeEntitiesFrom($entityList);
+        }
+
+        // Finally delete the service
+        $this->serviceRepository->delete($service);
+    }
+
+    /**
+     * Using the deleteCommandFactory, entity delete commands are created
+     * that will remove them from the appropriate environment.
+     */
+    private function removeEntitiesFrom(EntityList $entityList)
+    {
+        foreach ($entityList->getEntities() as $entity) {
+            try {
+                $command = $this->deleteCommandFactory->from($entity);
+                $this->commandBus->handle($command);
```

Comment:
I like the use of CommandHandlers very much because of it's clear structure. To inject a new Command from a CommandHandler into the bus again seems a bad practice because you don't expect at all that if you're working on the EntityDelete commands/handlers that they are initiated here.

Reply:
We discussed this IRL. I'm leaving it as is. The conclusion was: It would be best if the CommandHandlers would only perform tasks on a high level and use services to delegate the modifications on the domain. Now the command handlers are Jack of all trades (controller pattern ;) ) and it might be a good thing to refactor this in a future sprint.

I created a TD ticket to fix this in the future.